tapdb/burn_tree_test.go

```diff
@@ -0,0 +1,611 @@
+		)
+	})
+
+	// Test case 4: Inserting non-burn proof should fail.
+	t.Run("non_burn_proof", func(t *testing.T) {
+		// Create a non-burn leaf, the script key won't be an actual
+		// burn.
+		a := asset.RandAsset(t, asset.Normal)
+
+		op := test.RandOp(t)
+
+		nonBurnProof := &proof.Proof{
+			Asset:   *a,
+			PrevOut: op,
+		}
+
+		nonBurnLeaf := &universe.BurnLeaf{
+			UniverseKey: universe.LeafKey{
+				OutPoint:  op,
+				ScriptKey: &a.ScriptKey,
+			},
+			BurnProof: nonBurnProof,
+		}
+
+		result := burnTree.InsertBurns(ctx, spec, nonBurnLeaf)
+		require.Error(t, result.Err())
+		require.True(t, errors.Is(result.Err(), ErrNotBurn))
+	})
+
+	// Test case 5: Test idempotency - inserting the same burn leaf multiple
+	// times.
+	t.Run("idempotency", func(t *testing.T) {
+		// Create a new test environment to isolate the test.
+		burnTree, spec, burnLeaves := setupBurnTreeTest(t)
+
+		// Select a single burn leaf for the test.
+		burnLeaf := burnLeaves[0]
+
+		// Initial insertion.
+		result1 := burnTree.InsertBurns(ctx, spec, burnLeaf)
+		authLeaves1, err := result1.Unpack()
+		require.NoError(t, err)
+		require.Len(t, authLeaves1, 1)
+
+		// Store the root and proof from the first insertion.
+		originalRoot := authLeaves1[0].BurnTreeRoot
+
+		// Check sum after first insertion.
+		sumResult1 := burnTree.Sum(ctx, spec)
+		sumOpt1, err := sumResult1.Unpack()
+		require.NoError(t, err)
+
+		originalSum := sumOpt1.UnwrapOrFail(t)
+		require.Equal(t, burnLeaf.BurnProof.Asset.Amount, originalSum)
+
+		// Insert the same burn leaf again.
+		result2 := burnTree.InsertBurns(ctx, spec, burnLeaf)
+		authLeaves2, err := result2.Unpack()
+		require.NoError(t, err)
+		require.Len(t, authLeaves2, 1)
+
+		// The root should be the same as the original, indicating the
+		// tree didn't change.
+		require.Equal(
+			t, originalRoot.NodeHash(),
+			authLeaves2[0].BurnTreeRoot.NodeHash(),
+		)
+		require.Equal(
+			t, originalRoot.NodeSum(),
+			authLeaves2[0].BurnTreeRoot.NodeSum(),
+		)
+
+		// The sum should not change after inserting the same leaf
+		// again.
+		sumResult2 := burnTree.Sum(ctx, spec)
+		sumOpt2, err := sumResult2.Unpack()
+		require.NoError(t, err)
+
+		newSum := sumOpt2.UnwrapOrFail(t)
+		require.Equal(t, originalSum, newSum)
+
+		// Insert the same burn leaf a third time.
+		result3 := burnTree.InsertBurns(ctx, spec, burnLeaf)
+		require.NoError(t, result3.Err())
+
+		// Query for the burn leaf - should only find one instance.
+		queryResult := burnTree.QueryBurns(
+			ctx, spec, burnLeaf.BurnProof.OutPoint(),
+		)
+		queryOpt, err := queryResult.Unpack()
+		require.NoError(t, err)
+		queryLeaves := queryOpt.UnwrapOrFail(t)
+
+		// There should be exactly one leaf for this outpoint+scriptkey
+		// combination, not multiple.
+		count := 0
+		for _, leaf := range queryLeaves {
+			if leaf.BurnLeaf.UniverseKey.OutPoint ==
+				burnLeaf.UniverseKey.OutPoint &&
+				leaf.BurnLeaf.UniverseKey.ScriptKey.PubKey.IsEqual( //nolint:lll
```

Comment:
nit: can add `//nolint` directives above a block (e.g. above the `if`) so it counts for the whole section.